### PR TITLE
DPTS-510 Integration test pipeline

### PR DESCRIPTION
**Pull request by timrickwood**
See https://nhsd-confluence.digital.nhs.uk/display/DPTS/Integration+testing

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -482,14 +549,156 @@ create-artefact-repositories: # Create ECR repositories to store the artefacts -
...
+		make poll-s3-for-file MAX_ATTEMPTS=$(MAX_ATTEMPTS) BUCKET=$(BUCKET) FILENAME=$$filename
+	done
+
+load-single-integration-test-file-to-s3:  ### Upload single file to bucket - mandatory: FILENAME=[name of file],BUCKET=[name of folder in bucket]
+	eval "$$(make aws-assume-role-export-variables)"
+	path="test/integration/test-files"
+	make aws-s3-upload FILE=$$path/$(FILENAME) URI=$(BUCKET)/$(FILENAME)
+
+check-single-integration-test-file:## check if file has been archived mandatory: [MAX_ATTEMPTS]  BUCKET=[name of folder in bucket], FILENAME=[name of file]
+	# path="test/integration/test-files"
```
Comment (tamaragoldschmidt) [inline Makefile:650]: Should this comment be removed?

File: Makefile [hunk 2]
```diff
@@ -482,14 +549,156 @@ create-artefact-repositories: # Create ECR repositories to store the artefacts -
...
+# targets to test results of hk job
+#=================
+
+invoke-test-check: ###Run hk integration test lambda to check result of task - Mandatory [PROFILE] [TASK]
+	eval "$$(make aws-assume-role-export-variables)"
+	aws lambda invoke --function-name $(TF_VAR_hk_integration_tester_lambda_function_name) \
+	--payload '{ "task": "$(TASK)" }' \
+	/dev/stdout
+
+# lambda returns
```
Comment (tamaragoldschmidt) [inline Makefile:673]: Is this a descriptive comment on run-integration-test-lambda or comments during development?
If descriptive can you reformat

File: build/automation/var/profile/integration.mk [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ENVIRONMENT := integration
+SERVICE_PREFIX := $(PROJECT_ID)-$(ENV)
+VULNERABILITY_LEVEL := CRITICAL,HIGH
+
+# ==============================================================================
+# Infrastructure variables
+
+STACKS := secrets,s3,security-groups
+# TODO check full list ,symptomdiscriminators,stt,symptomgroups,symptomdiscriminators
+# don't need crons
```
Comment (tamaragoldschmidt) [inline build/automation/var/profile/integration.mk:17]: remoce comment

File: test/integration/models/referralrole.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+def create_referral_role_query(referral_role_ids):
+    query = (
+        """select id, name from pathwaysdos.referralroles where id in """+str(referral_role_ids)+""" order by id asc;"""
+    )
+    data = str(referral_role_ids)
+    return query, data
+
+
+# That a new record exists with id of 2000 and description of "Integration Test Create"
+# That an existing record with id of 2001 has updated description of "Integration Test Update"
```
Comment (tamaragoldschmidt) [inline test/integration/models/referralrole.py:40]: what do these 3 comments relate to?

File: test/unit/hk/integration/test_handler.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+# This may be hard to maintain as nothing in this list can be in its actual position
+# add any new task as penultimate in list
+def test_invalid_task_list():
+    temp_valid_tasks = ("symptomgroups","referralroles","data")
+    assert len(handler.valid_tasks) == len(temp_valid_tasks)
+    for i in range(len(handler.valid_tasks)):
+        assert handler.valid_tasks[i] != temp_valid_tasks[i]
+
+# TODO restore
```
Comment (tamaragoldschmidt) [inline test/unit/hk/integration/test_handler.py:44]: should this be removed or uncommented?

File: test/unit/hk/integration/test_handler.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+# @patch(f"{file_path}.symptomgroup.check_symptom_groups_data", return_value=True)
+# @patch(f"{file_path}.logger.log_for_audit")
+# @patch(f"{file_path}.run_data_checks_for_hk_task", return_value=True)
+# @patch(f"{file_path}.insert_test_data", return_value="")
+# @patch(f"{file_path}.database.close_connection", return_value="")
+# @patch(f"{file_path}.database.connect_to_database", return_value="db_connection")
+# def test_request_valid_hk_task(mock_db_connect, mock_db_close, mock_insert, mock_data_checks, mock_audit, mock_check):
+#     payload = generate_event_payload('symptomgroups')
+#     result = handler.request(event=payload, context=None)
+#     assert mock_db_connect.call_count == 1
```
Comment (tamaragoldschmidt) [inline test/unit/hk/integration/test_handler.py:154]: Should these 3 tests be removed or uncommented?

File: build/jenkins/Jenkinsfile.hk-integration [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+        DB_PREFIX = 'pathwaysdos'
+        DB_SUFFIX = 'tasks'
+        PIPELINE_DATABASE = "${DB_PREFIX}_${ENVIRONMENT}"
+        BUCKET = "uec-dos-tasks-${PROFILE}-housekeeping-bucket/${ENVIRONMENT}"
+        MAX_ATTEMPTS = 5
+        DB_HOST = 'uec-core-dos-pipeline-primary.dos-db-rds'
+        SLACK_CHANNEL = 'dos-tasks-pipeline-notifications-nonprod'
+        // TODO eventually develop
+        PATHWAYSDOS_V4_BRANCH = 'task%2FDPTS-811-Poc-int-test'
+        //TODO because we don't want to delete ns at start - not needed ?
```
Comment (tamaragoldschmidt) [inline build/jenkins/Jenkinsfile.hk-integration:48]: Should this TO_DO have been actioned?

File: build/jenkins/Jenkinsfile.hk-integration [hunk 8]
```diff
@@ -0,0 +1,190 @@
+pipeline {
+    agent any
+    parameters {
+        string(
+            description: 'Housekeeping task to run',
+            name: 'Task',
+            defaultValue: 'all'
+        )
+        string(
+            description: 'Test use tba',
```
Comment (tamaragoldschmidt) [inline build/jenkins/Jenkinsfile.hk-integration:10]: use tba.
What option other than integration could you have?

File: Makefile (hunk 1 above)
Comment (timrickwood) [reply Makefile:650]: yes

File: Makefile (hunk 2 above)
Comment (timrickwood) [reply Makefile:673]: reformatted

File: build/automation/var/profile/integration.mk (hunk 3 above)
Comment (timrickwood) [reply build/automation/var/profile/integration.mk:17]: will do

File: test/integration/models/referralrole.py (hunk 4 above)
Comment (timrickwood) [reply test/integration/models/referralrole.py:40]: seems to have been a duplicate - removed

File: test/unit/hk/integration/test_handler.py (hunk 6 above)
Comment (timrickwood) [reply test/unit/hk/integration/test_handler.py:154]: removed

File: build/jenkins/Jenkinsfile.hk-integration-test [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    environment {
+        BUILD_DATE = sh(returnStdout: true, script: "date -u +'%Y-%m-%dT%H:%M:%S%z'").trim()
+        // ENVIRONMENT = 'integration'
+        TASK = "${params.Task}"
+        BUCKET = "${params.Bucket}"
+        PROFILE = "${params.Profile}"
+        FILENAME = "int_${TASK}.csv"
+        // BUCKET = "uec-dos-tasks-${PROFILE}-housekeeping-bucket/${ENVIRONMENT}"
+        MAX_ATTEMPTS = 5
+        // DB_HOST = 'uec-core-dos-pipeline-primary.dos-db-rds'
```
[the remark this answers is not in the document]
Comment (timrickwood) [reply build/jenkins/Jenkinsfile.hk-integration-test:33]: removed

File: build/jenkins/Jenkinsfile.hk-integration [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+        NAMESPACE_RD = "${NAMESPACE}-rd"
+        NAMESPACE_CRON = "${NAMESPACE}-cron"
+        DB_PREFIX = 'pathwaysdos'
+        DB_SUFFIX = 'tasks'
+        PIPELINE_DATABASE = "${DB_PREFIX}_${ENVIRONMENT}"
+        BUCKET = "uec-dos-tasks-${PROFILE}-housekeeping-bucket/${ENVIRONMENT}"
+        MAX_ATTEMPTS = 5
+        DB_HOST = 'uec-core-dos-pipeline-primary.dos-db-rds'
+        SLACK_CHANNEL = 'dos-tasks-pipeline-notifications-nonprod'
+        // TODO eventually develop
```
[the remark this answers is not in the document]
Comment (timrickwood) [reply build/jenkins/Jenkinsfile.hk-integration:46]: no - it was more of a reminder - removed

File: build/jenkins/Jenkinsfile.hk-integration (hunk 7 above)
Comment (timrickwood) [reply build/jenkins/Jenkinsfile.hk-integration:48]: No again. I think this came across from cut-n-paste operation. Have removed

File: build/jenkins/Jenkinsfile.hk-integration [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+                script {
+                    echo 'Precautionary cleardown hk lambdas and integration test infrastructure'
+                    sh "make destroy PROFILE=${PROFILE} TASK=${params.Task}"
+                }
+            }
+        }
+        stage('Triggering build of database') {
+            steps {
+                echo "Running pathwaysdos_v4 pipeline ${DB_BUILD_JOB_NAME}"
+                // echo "Passing DELETE parameter: ${params.DELETE}"
```
[the remark this answers is not in the document]
Comment (timrickwood) [reply build/jenkins/Jenkinsfile.hk-integration:71]: will do

File: build/jenkins/Jenkinsfile.hk-integration [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+                    echo 'Teardown hk lambdas and integration test infrastructure'
+                    sh "make destroy PROFILE=${PROFILE} TASK=${params.Task}"
+                }
+            }
+        }
+        stage('Delete namespace and drop database') {
+            when { expression { params.Debug == false  && params.Task != "none" } }
+            steps {
+                echo "Running pathwaysdos_v4 pipeline ${DB_DROP_JOB_NAME}"
+                // parameters required to use existing delete job
```
[the remark this answers is not in the document]
Comment (timrickwood) [reply build/jenkins/Jenkinsfile.hk-integration:174]: no will remove

File: build/jenkins/Jenkinsfile.hk-integration (hunk 8 above)
Comment (timrickwood) [reply build/jenkins/Jenkinsfile.hk-integration:10]: good shout - I have improved description and in effect it is just integration

**Review by tamaragoldschmidt — APPROVED** [review]

**Pull request merged**